lutris/util/graphics/drivers.py

```diff
@@ -2,83 +2,139 @@
 
 Everything in this module should rely on /proc or /sys only, no executable calls
 """
+import contextlib
 import os
 import re
+from typing import Dict, Iterable, List
 
+from lutris.util.graphics.glxinfo import GlxInfo
 from lutris.util.log import logger
+from lutris.util.system import read_process_output
 
 MIN_RECOMMENDED_NVIDIA_DRIVER = 415
 
 
-def get_nvidia_driver_info():
+def get_nvidia_driver_info() -> Dict[str, Dict[str, str]]:
     """Return information about NVidia drivers"""
     version_file_path = "/proc/driver/nvidia/version"
     if not os.path.exists(version_file_path):
-        return
-    with open(version_file_path, encoding='utf-8') as version_file:
-        content = version_file.readlines()
-        nvrm_version = content[0].split(': ')[1].strip().split()
-        if 'Open' in nvrm_version:
+        return {}
+    with contextlib.suppress(OSError):
```

Comment:
My reason for choosing `contextlib.suppress` here was that this isn't so much catching an error per se as it is the fairly common-within-the-python-world exception handling as logic. In pseudocode, it's a pythonic version of:

```python
if can_read_proc_file():
  return use_proc_file()
else:
  return use_glxinfo()
```

I hesitate to use `logger.exception` (or, frankly, even `logger.warning`) here since this is a pretty normal condition. In addition to this happening in any snap-confined environment, this will happen in many other confined environments. In fact I'm quite surprised it's not occurring with NixOS, as their confinement is often stricter than what's provided for snaps. Stricter SELinux configurations will also block access to these files.

I'll put in a warning log for now, but I'm concerned this will result in logspam, especially since these functions will tend to be called together.

Reply:
Thank you.

Reply:
What about an info level log?

Reply:
If this is thought to be normal, that would be fine.

Reply:
OK, I've adjusted things as follows:

* Cases I specifically know could be expected (e.g. not having access to various `/proc` files) but that still allow normal usage: `info`
* Abnormal cases that still allow the same information to be retrieved (e.g. other `OSError` cases in `/proc/driver/nvidia/version`): `warning`
* Cases that could cause incorrect information to be provided: `error`

This should keep logspam to a minimum (All the cases I know of with either snap confinement or strict selinux rules should only provide `info` logs) without swallowing other issues (e.g. if `/proc/driver/nvidia/version` became a directory, it would still provide a warning).